File: app/routers/videos.py
from fastapi import APIRouter, Query

from ..db import connect, get_settings, typed_settings
from ..util import utc_now, ago_iso
from ..services import insights

router = APIRouter(prefix="/api/videos", tags=["videos"])

SORTS = {
    "published": lambda v: v["published_at"] or "",
    "views": lambda v: v["view_count"] or 0,
    "views_per_day": lambda v: v["views_per_day"] or 0,
    "delta": lambda v: v["view_delta"] or 0,
    "likes": lambda v: v["like_count"] or 0,
    "comments": lambda v: v["comment_count"] or 0,
    "ratio": lambda v: v["ratio"] or 0,
}


def query_videos(conn, group_id: int = 0, channel_id: int = 0, days: int = 30, kind: str = "all",
                 q: str = "", sort: str = "published", include_inactive: bool = False):
    now = utc_now()
    settings = typed_settings(get_settings(conn))
    where, params = insights.channel_where(group_id, include_inactive)
    if channel_id:
        where += " AND c.id = ?"
        params.append(channel_id)
    if days:
        where += " AND v.published_at >= ?"
        params.append(ago_iso(days=days, now=now))
    if kind == "short":
        where += " AND v.is_short = 1"
    elif kind == "long":
        where += " AND v.is_short = 0"
    if q.strip():
        where += " AND v.title LIKE ?"
        params.append(f"%{q.strip()}%")

    rows = conn.execute(
        f"""SELECT v.*, c.title AS channel_title, c.thumbnail_url AS channel_thumbnail, c.group_id, g.name AS group_name
            FROM videos v JOIN channels c ON c.id = v.channel_id JOIN groups g ON g.id = c.group_id
            WHERE {where} ORDER BY v.published_at DESC LIMIT 5000""",
        params,
    ).fetchall()
    channel_ids = list({row["channel_id"] for row in rows})
    medians = insights.channel_medians(conn, channel_ids, settings["video_stats_window_days"], now)

    items = []
    for row in rows:
        item = insights.video_dict(row, now)
        item["channel_title"] = row["channel_title"]
        item["channel_thumbnail"] = row["channel_thumbnail"]
        item["group_id"] = row["group_id"]
        item["group_name"] = row["group_name"]
        med = medians.get(row["channel_id"])
        item["ratio"] = round(item["views_per_day"] / med, 2) if med and item["views_per_day"] else None
        items.append(item)
    items.sort(key=SORTS.get(sort, SORTS["published"]), reverse=True)
    return items


@router.get("")
def list_videos(group_id: int = 0, channel_id: int = 0, days: int = Query(30, ge=0, le=3650),
                kind: str = "all", q: str = "", sort: str = "published",
                limit: int = Query(50, ge=1, le=500), offset: int = Query(0, ge=0), include_inactive: bool = False):
    with connect() as conn:
        items = query_videos(conn, group_id, channel_id, days, kind, q, sort, include_inactive)
    total = len(items)
    page = items[offset:offset + limit]
    return {"videos": page, "total": total, "limit": limit, "offset": offset,
            "short_count": sum(1 for v in items if v["is_short"]),
            "long_count": sum(1 for v in items if not v["is_short"])}

File: app/static/js/views/videos.js
// 영상 탐색: 전체 채널의 영상을 필터·정렬
import { h, svg, fmt, badge, emptyState, store, clear, toast, debounce } from '../ui.js';
import { videoRow } from './channel.js';

export const title = '영상';
const PERIODS = [[7, '7일'], [30, '30일'], [90, '90일'], [0, '전체']];
const KINDS = [['all', '전체'], ['short', '쇼츠'], ['long', '롱폼']];
const SORTS = [['published', '최신순'], ['views', '조회수'], ['views_per_day', '조회수/일'], ['delta', '최근 증가'], ['ratio', '평소 대비'], ['likes', '좋아요'], ['comments', '댓글']];
const LIMIT = 50;

export async function render(container, params, ctx) {
    const state = {
        groupId: store.get('vid.group', 0), channelId: 0, days: store.get('vid.days', 30), kind: store.get('vid.kind', 'all'),
        sort: store.get('vid.sort', 'published'), query: '', offset: 0, items: [], total: 0, shortCount: 0, longCount: 0, multiplier: 1.75,
    };
    ctx.setTitle('영상', '');
    const [groups, channelList, settings] = await Promise.all([ctx.groups(), ctx.api.get('/api/channels'), ctx.api.get('/api/settings')]);
    state.multiplier = settings.settings.outlier_multiplier;

    const filterRow = h('div', { class: 'filter-row' });
    const summary = h('div', { class: 'muted', style: { fontSize: '13px', marginBottom: '10px' } });
    const card = h('div', { class: 'card' });
    const more = h('div', { style: { textAlign: 'center', marginTop: '14px' } });
    container.append(filterRow, summary, card, more);

    function buildFilters() {
        clear(filterRow);
        const groupSelect = h('select', { class: 'select select-sm', id: 'vidGroup' }, h('option', { value: '0' }, '전체 그룹'),
            ...groups.map(g => h('option', { value: String(g.id), selected: g.id === state.groupId }, g.name)));
        groupSelect.addEventListener('change', () => { state.groupId = Number(groupSelect.value); store.set('vid.group', state.groupId); buildFilters(); reload(); });
        const channels = channelList.channels.filter(c => !state.groupId || c.group_id === state.groupId);
        const channelSelect = h('select', { class: 'select select-sm', id: 'vidChannel' }, h('option', { value: '0' }, '전체 채널'),
            ...channels.map(c => h('option', { value: String(c.id), selected: c.id === state.channelId }, c.title)));
        channelSelect.addEventListener('change', () => { state.channelId = Number(channelSelect.value); reload(); });

        const seg = (items, current, onPick) => {
            const el = h('div', { class: 'segment' });
            items.forEach(([value, label]) => {
                const btn = h('button', { type: 'button', class: value === current ? 'active' : '' }, label);
                btn.addEventListener('click', () => { onPick(value); el.querySelectorAll('button').forEach(b => b.classList.toggle('active', b === btn)); });
                el.appendChild(btn);
            });
            return el;
        };
        const sortSelect = h('select', { class: 'select select-sm', id: 'vidSort' }, ...SORTS.map(([v, l]) => h('option', { value: v, selected: v === state.sort }, l)));
        sortSelect.addEventListener('change', () => { state.sort = sortSelect.value; store.set('vid.sort', state.sort); reload(); });
        const search = h('input', { class: 'input input-sm', type: 'search', placeholder: '제목 검색', id: 'vidSearch', value: state.query });
        search.addEventListener('input', debounce(() => { state.query = search.value.trim(); reload(); }, 250));
        const exportBtn = h('a', { class: 'btn btn-secondary btn-sm', href: '#' }, svg('download'), 'CSV');
        exportBtn.addEventListener('click', (e) => { e.preventDefault(); window.location.href = `/api/export/videos.csv?${queryString()}`; });

        filterRow.append(
            h('div', { class: 'field' }, h('span', { class: 'field-label' }, '게시 기간'), seg(PERIODS, state.days, v => { state.days = v; store.set('vid.days', v); reload(); })),
            h('div', { class: 'field' }, h('span', { class: 'field-label' }, '유형'), seg(KINDS, state.kind, v => { state.kind = v; store.set('vid.kind', v); reload(); })),
            h('div', { class: 'field' }, h('label', { for: 'vidGroup' }, '그룹'), groupSelect),
            h('div', { class: 'field' }, h('label', { for: 'vidChannel' }, '채널'), channelSelect),
            h('div', { class: 'field' }, h('label', { for: 'vidSort' }, '정렬'), sortSelect),
            h('div', { class: 'field grow' }, h('label', { for: 'vidSearch' }, '검색'), h('div', { class: 'search-box' }, svg('search'), search)),
            h('div', { class: 'field' }, h('span', { class: 'field-label' }, ' '), exportBtn),
        );
    }

    function queryString() {
        return new URLSearchParams({ group_id: state.groupId, channel_id: state.channelId, days: state.days, kind: state.kind, sort: state.sort, q: state.query }).toString();
    }

    async function reload() { state.offset = 0; state.items = []; await load(); }

    async function load() {
        container.classList.add('is-loading');
        try {
            const r = await ctx.api.get('/api/videos', { group_id: state.groupId, channel_id: state.channelId, days: state.days, kind: state.kind,
                                                       sort: state.sort, q: state.query, limit: LIMIT, offset: state.offset });
            state.items = state.offset ? state.items.concat(r.videos) : r.videos;
            state.total = r.total; state.shortCount = r.short_count; state.longCount = r.long_count;
            renderTable();
        } catch (error) { toast(error.message, 'error'); }
        finally { container.classList.remove('is-loading'); }
    }

    function renderTable() {
        clear(card); clear(more);
        summary.textContent = `${fmt.int(state.total)}개 영상 · 쇼츠 ${fmt.int(state.shortCount)} · 롱폼 ${fmt.int(state.longCount)}`;
        ctx.setTitle('영상', summary.textContent);
        if (!state.items.length) {
            card.appendChild(emptyState({ iconName: 'video', title: '조건에 맞는 영상이 없습니다', text: '기간이나 유형 필터를 넓혀 보세요. 영상은 채널 업데이트 때 수집됩니다.' }));
            return;
        }
        const table = h('table', { class: 'table' }, h('thead', {}, h('tr', {},
            h('th', { class: 'sticky-col' }, '영상'), h('th', {}, '게시'), h('th', { class: 'num' }, '조회수'), h('th', { class: 'num' }, '조회수/일'), h('th', { class: 'num' }, '최근 증가'),
            h('th', { class: 'num' }, '좋아요'), h('th', { class: 'num' }, '댓글'), h('th', {}, '길이'), h('th', { class: 'num' }, '평소 대비'))));
        const tbody = h('tbody');
        state.items.forEach(v => tbody.appendChild(videoRow(v, state.multiplier, h('a', { href: `#/channels/${v.channel_id}` }, v.channel_title))));
        table.appendChild(tbody);
        card.appendChild(h('div', { class: 'table-wrap' }, table));
        if (state.items.length < state.total) {
            more.appendChild(h('button', { class: 'btn btn-secondary', type: 'button', onclick: () => { state.offset += LIMIT; load(); } },
                `더 보기 (${fmt.int(state.items.length)} / ${fmt.int(state.total)})`));
        }
    }

    buildFilters();
    await load();
    const off = ctx.onRefreshDone(() => reload());
    return () => off();
}
